eras/shelley/impl/src/Cardano/Ledger/Shelley/AdaPots.hs

```diff
@@ -72,6 +93,87 @@ totalAdaES cs =
         reservesAdaPot,
         rewardsAdaPot,
         utxoAdaPot,
+        -- keyDepositAdaPot,  -- We don't count these two, as their
+        -- poolDepositAdaPot, -- sum is always depositsAdaPot
         depositsAdaPot,
         feesAdaPot
       } = totalAdaPotsES cs
+
+-- =============================================
+-- Produced and Consumed are specialized AdaPots
+-- relative to the actions of a TxBody
+
+-- | Itemizing what is consumed by a transaction
+data Consumed = Consumed
+  {conInputs :: !Coin, conRefunds :: !Coin, conWithdrawals :: !Coin}
+
+instance Show Consumed where
+  show (Consumed (Coin i) (Coin r) (Coin w)) =
+    "Consumed(Inputs " ++ show i ++ ", Refunds " ++ show r ++ ", Withdrawals " ++ show w ++ ") = " ++ show (i + r + w)
+
+-- | Itemizing what is Produced by a transaction
+data Produced = Produced
+  {proOutputs :: !Coin, proFees :: Coin, proDeposits :: !Coin}
+
+instance Show Produced where
+  show (Produced (Coin out) (Coin f) (Coin d)) =
+    "Produced(Outputs " ++ show out ++ ", Fees " ++ show f ++ " Deposits " ++ show d ++ ") = " ++ show (out + f + d)
+
+-- =========================
+
+-- | Compute the Coin part of what is consumed by a TxBody, itemized as a 'Consume'
+consumedTxBody ::
```

Comment:
Seems surprising to have these two  data types `Consumed` and `Produced` as well as the functions, `consumedTxBody` and `producedTxBody` (which are functions of a transaction)  here in `AdaPots`.   What is their relation with the rest of this module ?

Reply:
These are just another kind of Ada Pot, they are the Ada Pots induced by a transaction. We often get errors like produced is not equal to consumed, and the give us two coins, but without telling how those coins were computed. So this makes that explicit.